### PR TITLE
feat: implement support for data file uploads

### PR DESCRIPTION
**Pull request by IlliaDerevianko**
UI:
  - Add a button to open the dialog with CSV upload
  - Add a CSV upload dialog with a dropdown for the upload type
  - Display error/success message depending on the outcome of the upload

Server:
  - Add a server action that parses CSV, updates the appropriate db table (only User so far) and returns a message

Screenshot 2024-09-10 at 15 50 17
*Homepage with the upload button*

Screenshot 2024-09-12 at 13 38 47
*CSV upload dialog with the drag and drop area*

### CODE REVIEW

**Review by jasebaileyicl — COMMENTED** [review]
I've had a look and added some comments

File: components/DataUpload.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          <Flex direction="column" gap="3">
+            <LabelledInput label="Data regarding">
+              <Dropdown
+                values={Object.keys(DataUploadEnum)}
+                currentValue={dataUploadChoice}
+                onValueChange={(value) => setDataUploadChoice(value as DataUploadEnum)}
+              />
+            </LabelledInput>
+
+            <Dropzone onDrop={(acceptedFiles) => setFile(acceptedFiles[0])} maxFiles={1}>
```
[the remark this answers is not in the document]
Comment (IlliaDerevianko) [reply components/DataUpload.tsx:53]: Just updated the screenshot

File: lib/types.ts [hunk 2]
```diff
@@ -7,3 +7,15 @@ export enum NextActionEnum {
   INFORM_CANDIDATE,
   CANDIDATE_INFORMED
 }
+
+export enum DataUploadEnum {
```
[the remark this answers is not in the document]
Comment (zaki-amin) [reply lib/types.ts:11]: We'll add some tooltips or a description to clarify what each upload does. Course is just a placeholder for now, it'll be changed probably to _Outcome_ which will allow uploads for the outcome for each course applied to.

File: lib/upload.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  role: RoleEnum
+})
+
+const applicantInsertSchema = z.object({
+  cid: z.number().int().positive()
+})
+
+const courseInsertSchema = z.object({
+  degreeCode: z.string()
+})
```
[the remark this answers is not in the document]
Comment (zaki-amin) [reply lib/upload.ts:24]: This is also a placeholder for uploading outcomes per course applied for as explained in the previous code. The codes are given as G700.1 for Computing (Artificial Intelligence and Machine Learning) (MEng 4YFT), not entirely sure where this comes from?

**Comment by zaki-amin** [comment]
This PR is now out of date given that most of the logic in `upload.ts` has been improved on another branch. Furthermore, the example data provided by Ritu has suggested a new approach in which applicants and applications are created together, so separate upserts no longer make sense. All the upload functionality will be submitted in a different PR.

**Pull request closed without merging**